Eppo-exp/vercel-edge
# Pavel/vercel edge sdk client

This is Eppo SDK configured to work with [Vercel](https://vercel.com/)

### Description

This is quite an interesting sdk, as it is neither client nor node.js sdk.
Customer requested for an SDK, which can work in [Vercel Edge Middlewares](https://vercel.com/docs/functions/edge-middleware)

While Next.js, for example, has the same middlewares running, they do not have [limitations of vercel edge middlewares](https://vercel.com/docs/functions/edge-middleware/limitations)

From the official docs:

```
[Key takeaways for Edge Middleware on Next.js](https://vercel.com/docs/functions/edge-middleware/quickstart#key-takeaways-for-edge-middleware-on-next.js)
Edge Middleware runs before the cache
You can import helpers that extend Web API objects (NextResponse, NextRequest, see [Edge Middleware API](https://vercel.com/docs/functions/edge-middleware/middleware-api) for more information on these APIs)
You can use a custom matcher config to only trigger the middleware in specific routes
**You cannot use Node.js APIs in Edge Middleware**
```

This is why this JS SDK does not have:
- polling;
- LRUCache

Not having polling is fine, as we do the same in js-client-sdk - fetch config once on init and store it to some storage (in client it is localStorage). 
For Vercel, as it is not browser env, it is not possible to use localStorage, however, it is possible to use [Vercel Edge Config Store](https://vercel.com/docs/storage/edge-config)
This is Vercel's solution for storage, with fast reads and slow writes.
Basically, adapter for it is the main thing developed here.

Example of Vercel Middleware:

```ts
import { NextResponse, NextRequest } from 'next/server';

import { init } from './services/eppo-client';
import { IAssignmentLogger } from '@eppo/js-client-sdk-common';

const assignmentLogger: IAssignmentLogger = {
  logAssignment(assignment) {
    console.log('assignement', assignment)
  },
};

export async function middleware(request: NextRequest) {
  try {
    if (!process.env.EDGE_CONFIG) {
      throw new Error('Define EDGE_CONFIG env variable');
    }

    if (!process.env.EDGE_CONFIG_STORE_ID) {
      throw new Error('Define EDGE_CONFIG_STORE_ID env variable')
    }

    if (!process.env.EDGE_CONFIG_TOKEN) {
      throw new Error('Define EDGE_CONFIG_TOKEN env variable')
    }

    if (!process.env.INTERNAL_FEATURE_FLAG_API_KEY) {
      throw new Error('Define INTERNAL_FEATURE_FLAG_API_KEY env variable')
    }

    const user = getCurrentUser();

    const eppoClient = await init({
      apiKey: process.env.INTERNAL_FEATURE_FLAG_API_KEY,
      assignmentLogger,
      vercelParams: {
        edgeConfig: process.env.EDGE_CONFIG,
        edgeConfigStoreId: process.env.EDGE_CONFIG_STORE_ID,
        vercelApiToken: process.env.EDGE_CONFIG_TOKEN,
      }
    });

    const variation = eppoClient.getBooleanAssignment('show-new-feature', user.id, { 
      'country': user.country,
      'device': user.device,
    }, false);

    console.log('show-new-feature enabled:', variation);

   const response =  NextResponse.next();

    response.headers.set('show-new-feature', isEnabled ? 'enabled': 'disabled');

    return response;
  } catch (error) {
    console.error('Error in middleware:', error);
    
    // Return an error response if needed
    return NextResponse.error();
  }
}

// Define paths where middleware should run
export const config = {
  matcher: [
    '/((?!api|_next/static|_next/image|favicon.ico|.*\\.js$|.*\\.css$|.*\\.map$|.*\\.json$|.*\\.png$|.*\\.jpg$|.*\\.jpeg$|.*\\.gif$|.*\\.svg$).*)',
  ],
};

```


One thing I don't like - is that there is no Vercel package to write to Vercel Edge Config.
Because of this, I need to combine usage of Vercel API to write data to store and vercel-client to read data from store.

### How to test

Well, i tested this with Vercel.

1) Create account at Vercel using github account;
2) Created a new repo and added a vercel integration to it;
3) created a project out of next.js template;
4) checked out newly created and populated with next.js repo to local;
5) created a Vercel Edge Config Store
6) Setup required envs in vercel
7) created a middleware as above (SDK is not available in npm yet, so had to copy its code to the repo)
8) push to main (or any other branch) starts Vercel Deployment
9) Go to a deployed website and check response headers
